test/unit/mt_membar.c

```diff
@@ -0,0 +1,189 @@
+#include <stdio.h>
+#include <assert.h>
+#include <pthread.h>
+#include <shmem.h>
+
+/* For systems without the PThread barrier API (e.g. MacOS) */
+#include "pthread_barrier.h"
+
+#define T 2
+#define ITER 100
+
+#ifndef MAX
+#define MAX(A,B)   (((A)>(B)) ? (A) : (B))
+#endif
+
+int shared_dest_1 = 0, shared_dest_2 = 0, result = 0;
+int me, npes, errors = 0, sum_error = 0;
+long lock = 0;
+
+pthread_barrier_t fencebar;
+
+long pSync[_SHMEM_REDUCE_SYNC_SIZE];
+int pWrk[MAX(1, _SHMEM_REDUCE_MIN_WRKDATA_SIZE)];
+
+static void * thread_main(void *arg) {
+    int tid = *(int *) arg;
+    int one = 1, zero = 0;
+    int i;
+
+    /* TEST WAIT */
+    for (i = 0; i < ITER; i++) {
+        if (tid == 0) {
+            shared_dest_1 = 1;
+            shmem_int_wait_until(&shared_dest_1, SHMEM_CMP_EQ, zero);
+            shmem_int_atomic_add(&result, one, me);
```

Comment:
Can we replace this with:
```
result++;
shmem_fence();
```
IIUC, we should have atomicity from the "token" that is being passed around via `shared_dest_1`.  We just need to flush writes before passing the token, which is what `shmem_fence` is supposed to handle.

Reply:
Note this substitution is for both threads 1 and 2.

Reply:
Ok -- makes sense.  Do you think this would be potentially useful as another test?
```
shmem_int_wait_until(&shared_dest_1, SHMEM_CMP_EQ, tid);
result++;
shmem_fence();
shared_dest_1 = (tid+1) % T;
```